Allowed format/characters for member names

cziegenberg opened the pull request:
This patch contains the changes discussed in #567, to reach the following goals:
- Define basic MUSTs for naming members that were missing in the spec (including support for NON-ASCII characters)
- Give API designers the freedom to use another naming scheme if required
- Define recommendations to standardize the naming

This suggestion includes several considerations
- Existing naming of reserved members
- URL mapping / URL safe characters ([RFC 3986](http://tools.ietf.org/html/rfc3986))
- Readability and usability, common spelling
- Reduction of potential implementation problems with some languages
- Not breaking the previous naming recommendations

cziegenberg committed:
Added member names section

Added member names section to define the allowed characters and naming rules for member names

cziegenberg committed:
Changed recommended seperator character

Changed recommended separator character, because it could cause problems in some implementation languages and to follow the added member name format recommendations in the spec.

cziegenberg committed:
Changed missing word

cziegenberg committed:
Optimized SHOULDs for the member naming

- Prefer minus over underscore (after discussion about that)
- Optimized recommended characters to URL safe ones
- Minor changes

cziegenberg committed:
Implemented discussed changes

dgeb commented:
@ziege thanks for your work here and in #567.

I believe that many of these restrictions are sound but some go too far. It is important to allow for public discussion of each restriction. If you don't mind making a lot of tweaks, we can comment on them inline.

To start, I'd prefer that the recommendations marked **SHOULD** be moved exclusively to the recommendations page.

Also, there needs to be a clarification of the term ASCII. We are discussing non-control characters 32-126 only, correct? I believe that extended ASCII codes (>= 128) should probably be covered in the "non-ASCII" category.

cziegenberg commented:
> If you don't mind making a lot of tweaks, we can comment on them inline.

Let's try it...

> I'd prefer that the recommendations marked SHOULD be moved exclusively to the recommendations page.

I also thought about placing it there, but didn't because these recommendations are relatively important. I also thought about linking the recommendation page but wanted to keep the relevant information in one document.

> We are discussing non-control characters 32-126 only, correct?

Yes, the allowed ASCII characters are defined in detail in the sub sections of the added "Member names" section.

ziege commented:
Sorry, one big comment... Didn't work on my mobile.

dgeb commented:
> Yes, the allowed ASCII characters are defined in detail in the sub sections of the added "Member names" section.

The "Member names" section does not define what is meant by ASCII characters. It calls out specific ASCII characters, but does not define the whole set. Therefore, it is also unclear what is meant by "non-ASCII" characters.

dgeb commented:
> I also thought about placing it there, but didn't because these recommendations are relatively important. I also thought about linking the recommendation page but wanted to keep the relevant information in one document.

To tell you the truth, I am reluctant to even include these as recommendations. Although the specification is in English and uses English reserved keys, non-English characters are perfectly valid for implementation-specific keys.

cziegenberg commented:
Ah, okay. First we called it UNICODE characters, but as this can overlap with ASCII, I changed it to NON-ASCII...

I will change it. Perhaps "UNICODE except ASCII (code 0 - 127)" is clearer. Or do you have a better suggestion?

dgeb commented:
Although it's a little awkward, I'm not opposed to the term "non-ASCII", but believe it needs to be defined. I think that "UNICODE except ASCII (code 0 - 127)" is the right definition.

cziegenberg committed:
Implemented discussed changes

- Clarification of ASCII/NON-ASCII/UNICODE
- Removed recommendations that are not relevant for URL mapping (will be added to separate recommendations page)

cziegenberg committed:
Implemented discussed changes

- Avoided the word ASCII, as it's not clearly defined.
- Moved recommendations to standardize member names here.
- URL mapping recommendations are defined in the format spec.

cziegenberg commented:
I did the following changes:
- Removed the word "ASCII" as it wasn't clearly defined
- Divided the recommendations in two parts:
  - General recommendation in the format spec to use URL safe characters, which is relevant for URL mapping.
  - Further recommendations to standardize the member name format in the recommendations page only.
- Avoided duplications

passcod commented on recommendations/index.md:
```diff
@@ -9,13 +9,14 @@ are beyond the scope of the base JSON API specification.
+
+Together with the recommendations in the format spec this results in:
+- Member names **SHOULD** start and end with the characters "a-z" (U+0061 to U+007A)
+- Member names **SHOULD** contain only the characters "a-z" (U+0061 to U+007A), "0-9" (U+0030 to U+0039) and "-" (U+002D HYPHEN-MINUS)
```
Why not just discard all that is in this PR and specify only those two rules, then? It seems simpler and much clearer (the long lists in the spec especially are really tedious).

cziegenberg committed:
Optimized recommendations

Optimized recommendations to avoid unnecessary duplication

cziegenberg commented:
@passcod: I reduced it to the two resulting rules.

ethanresnick commented:
@ziege Thanks for fighting for all these changes! This PR really looks great, at least from where I'm standing. In particular, reserving the punctuation that we're already using (for sorting etc.) seems so crucial that I can't believe it didn't come up before, and reserving other punctuation will, I suspect, be very helpful down the line for extending the query parameter syntax. (In particular, I think it'll be important for allowing us to reference new members at the top-level of a resource object, besides `type` and `id`.) So again, just great work!

cziegenberg commented:
Thanks for the positive feedback @ethanresnick!

tkellen commented:
:+1:

dgeb commented:
Thanks again very much, @ziege! :shipit:

cziegenberg commented:
Great! Thanks to you too!

jonkoops commented:
Thanks @ziege, these are some important additions.
